### PR TITLE
Enable Deployment updates

### PR DESCRIPTION
**Pull request by zaneb**
Unblock updating of the metal3 Deployment created by the CBO (but not those originally created by the MAO - this will be added in future by #65).

### CODE REVIEW

File: provisioning/baremetal_pod.go [hunk 1]
```diff
@@ -471,7 +471,9 @@ func NewMetal3Deployment(targetNamespace string, images *Images, config *metal3i
 func GetExistingMetal3Deployment(client appsclientv1.DeploymentsGetter, targetNamespace string) (bool, error) {
 	existing, err := client.Deployments(targetNamespace).Get(context.Background(), baremetalDeploymentName, metav1.GetOptions{})
 	if existing != nil && err == nil {
-		return true, nil
+		if _, maoOwned := existing.Annotations["machine.openshift.io/owned"]; maoOwned {
+			return true, nil
+		}
```
Comment (asalkeld) [inline provisioning/baremetal_pod.go:476]: i think this changes the entire meaning of the function from does it exist, to is the deployment owned by mao
maybe keep it out of the function?

File: controllers/provisioning_controller.go [hunk 2]
```diff
@@ -222,7 +219,7 @@ func (r *ProvisioningReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error
 		return ctrl.Result{}, errors.Wrap(err, "failed to create Inspector password")
 	}
 
-	// If Metal3 Deployment already exists, do nothing.
+	// If Metal3 Deployment already exists and managed by MAO, do nothing.
```
Comment (asalkeld) [inline controllers/provisioning_controller.go:222]: I think we need a function named ExistingDeploymentNotOwedByUS() - also get rid of the "Get"

File: controllers/provisioning_controller.go [hunk 3]
```diff
@@ -237,35 +234,59 @@ func (r *ProvisioningReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error
...
+	}
+
+	// Proceed with creating or updating the Metal3 deployment
+	updated, err := r.ensureMetal3Deployment(baremetalConfig, &containerImages)
 	if err != nil {
-		return ctrl.Result{}, fmt.Errorf("unable to put %q ClusterOperator in Syncing state: %v", clusterOperatorName, err)
+		return ctrl.Result{}, err
+	}
+	if updated {
+		return ctrl.Result{Requeue: true}, err
```
Comment (asalkeld) [inline controllers/provisioning_controller.go:251]: why is this here? this makes the function logically harder to parse.

File: controllers/provisioning_controller.go [hunk 4]
```diff
@@ -237,35 +234,59 @@ func (r *ProvisioningReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error
...
 	}
 
-	// Proceed with creating a new Metal3 deployment
-	metal3Deployment := provisioning.NewMetal3Deployment(ComponentNamespace, &containerImages, &baremetalConfig.Spec)
-	expectedGeneration := resourcemerge.ExpectedDeploymentGeneration(metal3Deployment, r.Generations)
+	if specChanged {
+		baremetalConfig.Status.ObservedGeneration = baremetalConfig.Generation
+		err = r.Client.Status().Update(context.Background(), baremetalConfig)
+		return ctrl.Result{Requeue: true}, err
```
Comment (asalkeld) [inline controllers/provisioning_controller.go:257]: also this, why don't we continue with the function

File: controllers/provisioning_controller.go [hunk 5]
```diff
@@ -237,35 +234,59 @@ func (r *ProvisioningReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error
...
+	err = r.updateCOStatus(ReasonComplete, "new Metal3 deployment completed", "")
 	if err != nil {
-		return ctrl.Result{}, fmt.Errorf("unable to set controllerReference on deployment: %v", err)
+		return ctrl.Result{}, fmt.Errorf("unable to put %q ClusterOperator in Available state: %v", clusterOperatorName, err)
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *ProvisioningReconciler) ensureMetal3Deployment(provConfig *metal3iov1alpha1.Provisioning, images *provisioning.Images) (updated bool, err error) {
```
Comment (asalkeld) [inline controllers/provisioning_controller.go:268]: nit, not a fan of named return variables and the associated empty returns

**Review by asalkeld — COMMENTED** [review]
left some comments..
I thinks this Reconcile function needs breaking up, it's getting to big.. (not your fault, it's just growing)

File: controllers/provisioning_controller.go (hunk 2 above)
Comment (zaneb) [reply controllers/provisioning_controller.go:222]: It's going to disappear soon anyway, but I agree this isn't very well named.

File: provisioning/baremetal_pod.go (hunk 1 above)
Comment (zaneb) [reply provisioning/baremetal_pod.go:476]: That does sounds less hacky, but I think we have to follow the plan because we need the MAO to do the right thing during downgrades, and the thing we taught it to do in 4.6 is look at the annotations. As far as I know the MAO doesn't update the Deployment at all once it exists, so I doubt that it rewrites the ControllerRef on downgrade, even if it does set it on create.
Again, this code is going away soon anyway. I _think_ we could actually just delete it now and it would work, since we call ApplyDeployment() and IIUC that should rewrite the annotations. But Sandhya has improved tests and stuff in #65, so I'm leaving it for that PR.

File: controllers/provisioning_controller.go (hunk 3 above)
Comment (zaneb) [reply controllers/provisioning_controller.go:251]: We wrote to the status, so we should requeue otherwise the next write to the status will fail.

File: controllers/provisioning_controller.go (hunk 4 above)
Comment (zaneb) [reply controllers/provisioning_controller.go:257]: We probably could in this case. I think I wrote this before I completely realised that the Generation only changes when the Spec is modified, so in fact there's no danger that writing the Status could cause us to bounce the CO state from Syncing->Complete->Syncing->Complete.

File: controllers/provisioning_controller.go (hunk 5 above)
Comment (zaneb) [reply controllers/provisioning_controller.go:268]: Yeah, I can't decide how I feel about them. In this case it seems essential to name the parameter to document what it means, and then after that it feels weird to me to not use the name in the function.

File: controllers/provisioning_controller.go (hunk 3 above)
Comment (asalkeld) [reply controllers/provisioning_controller.go:251]: seems silly we can't write to the status twice

File: controllers/provisioning_controller.go (hunk 5 above)
Comment (asalkeld) [reply controllers/provisioning_controller.go:270]: you are allowed to write a comment :smiley:

File: controllers/provisioning_controller.go (hunk 2 above)
Comment (sadasu) [reply controllers/provisioning_controller.go:222]: The logic around this is getting refined as part of https://github.com/openshift/cluster-baremetal-operator/pull/65. And, please feel free to propose better names.

File: provisioning/baremetal_pod.go (hunk 1 above)
Comment (sadasu) [reply provisioning/baremetal_pod.go:476]: MAO adds the maoOwnedAnnotations to all the resources it creates and so we just using it.

This documentation about ControllerRef says it is partially implemented https://stupefied-goodall-e282f7.netlify.app/contributors/design-proposals/api-machinery/controller-ref/. 

This is hand-off mechanism is required only during upgrade/downgrade between 4.6 and 4.7.

File: provisioning/baremetal_pod.go [hunk 6]
```diff
@@ -468,10 +468,12 @@ func NewMetal3Deployment(targetNamespace string, images *Images, config *metal3i
 	}
 }
 
-func GetExistingMetal3Deployment(client appsclientv1.DeploymentsGetter, targetNamespace string) (bool, error) {
+func MAOMetal3DeploymentExists(client appsclientv1.DeploymentsGetter, targetNamespace string) (bool, error) {
 	existing, err := client.Deployments(targetNamespace).Get(context.Background(), baremetalDeploymentName, metav1.GetOptions{})
 	if existing != nil && err == nil {
-		return true, nil
+		if _, maoOwned := existing.Annotations["machine.openshift.io/owned"]; maoOwned {
```
Comment (sadasu) [inline provisioning/baremetal_pod.go:474]: This check really is not necessary because even if the metal3 deployment was created by MAO, we are allowed to change it.

File: provisioning/baremetal_pod.go (hunk 1 above)
Comment (sadasu) [reply provisioning/baremetal_pod.go:476]: I do not think the name suggested here is correct. We want it to be a generic method to see if a metal3 deployment already exists after/during an upgrade. When we upgrade from 4.6 to 4.7, the older metal3 deployment would have been created by MAO. From 4.7 to 4.8, it would have been created by CBO. 
So, I am not for adding MAO in the name of the method.

**Comment by sadasu** [comment]
I think commit "Allow metal3 Deployment to be updated" is not necessary and comments against code in that commit reflect the reasons why. All other commits look good.
Also, as previously discussed #65 refactors a lot of the code around metal3 deployment better.

File: controllers/provisioning_controller.go (hunk 3 above)
Comment (sadasu) [reply controllers/provisioning_controller.go:251]: Can we move ensureMetal3Deployment() to baremetal_pod.go?

**Comment by sadasu** [comment]
> I think commit "Allow metal3 Deployment to be updated" is not necessary and comments against code in that commit reflect the reasons why. All other commits look good.
> Also, as previously discussed #65 refactors a lot of the code around metal3 deployment better.

Reading the description again, I see that the "Allow metal3 Deployment to be updated" commit is just meant to get you unblocked for further work. I am Ok for this to merge as is. I can rebase #65 over these changes and rework the logic around annotations to its final form.

**Comment by sadasu** [comment]
Nice changes to the expected generations code.

File: provisioning/baremetal_pod.go (hunk 1 above)
Comment (zaneb) [reply provisioning/baremetal_pod.go:476]: I am expecting that this method will either go away or be completely rewritten (with a new name) in #65.
